DDF-2515 Updated catalog-core-commands that support the --cql option to also support temporal options

emmberk opened the pull request:
#### What does this PR do?
This PR updates catalog-core-commands to new karaf command annotations
and updates catalog-core-commands that support the `--cql` option to also support temporal options. 
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@peterhuffer @glenhein @mweser
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@millerw8
@kcwire
#### How should this be tested?

1. Confirm unit tests succeed.
2. Confirm that the command still works as expected and that the `--help` message is displayed correctly for all of the catalog commands:

  - `catalog:describe`
  - `catalog:dump`
  - `catalog:envlist`
  - `catalog:ingest`
  - `catalog:inspect`
  - `catalog:latest`
  - `catalog:migrate`
  - `catalog:range`
  - `catalog:remove`
  - `catalog:removeall`
  - `catalog:replicate`
  - `catalog:search`
  - `catalog:seed`
  - `catalog:spatial` (see [DDF-1282](https://jira.di2e.net/browse/DDF-1282))
  - `catalog:validate`

3. Confirm that the following commands include the `--cql` and temporal options and that the filtering options work as described, and try to use different temporal options:

  - `catalog:dump`
  - `catalog:migrate`
  - `catalog:remove`
  - `catalog:replicate`
  - `catalog:search`
  - `catalog:seed` (see https://github.com/codice/ddf/pull/1054)

#### Any background context you want to provide?
Related to https://github.com/codice/ddf/pull/1203.

When reviewing the code, it would probably help to look at code changes from each of the two commits separately so that file renames are recognized correctly.
#### What are the relevant tickets?
[DDF-2515](https://codice.atlassian.net/browse/DDF-2515)
#### Screenshots (if appropriate)
#### Checklist:
- [x] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

rzwiefel commented on catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/CqlCommands.java:
```diff
@@ -0,0 +1,225 @@
+            + " Smaller --lastXXXX time units take precedence over larger time units.";
+
+    private static final String[] TEMPORAL_PROPERTIES =
+            {Core.CREATED, Metacard.EFFECTIVE, Core.EXPIRATION, Core.MODIFIED};
```
Any reason these specific properties were chosen? Do we not want to let them do any attribute that is a date/time type?

peterhuffer replied:
@rzwiefel These were added in a separate ticket, and previously only Metacard.EFFECTIVE was supported. https://github.com/codice/ddf/pull/1203

emmberk replied:
These properties are from https://github.com/codice/ddf/pull/1203. Would it be worthwhile to add every property that is date/time type?

peterhuffer replied:
It would be nice to do, but I would also prefer it to be a separate ticket.

rzwiefel replied:
We could do it dynamically. Pulling in the reference-list of metacard types and looping through the attribute descriptors for any with the `DATE_TYPE`. That would pull any Date types as they change.

In blueprint it would just be a `<reference-list interface="ddf.catalog.data.MetacardType" />` but since karaf commands use the annotations not sure if you can just use `@Reference` or if they have some sort of `reference-list` annotation. Then iterating through that list and calling `MetacardType::getAttributeDescriptors` and pulling out any with the date type should get us all the system date types dynamically.

glenhein commented on catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/CatalogCommands.java:
```diff
@@ -78,14 +93,19 @@ protected SolrCacheMBean getCacheProxy()
 
-    protected FilterBuilder getFilterBuilder() throws InterruptedException {
-        return getService(FilterBuilder.class);
+    protected <T> T getServiceByFilter(Class<T> clazz, String filter)
```
I like returning an Optional in cases where the code could otherwise return a null value.

peterhuffer commented on catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/CqlCommands.java:
```diff
@@ -0,0 +1,225 @@
+    int lastWeeks;
+
+    @Option(name = "--lastMonths", required = false, aliases = {"-m",
+            "-months"}, multiValued = false, description = "Option to query the last N month."
```
typo: `months.`

glenhein commented on catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/SeedCommand.java:
```diff
@@ -91,50 +83,21 @@
 
+    private final Predicate<Metacard> isNonCachedResource = metacard -> {
+        Attribute cached = metacard.getAttribute(RESOURCE_CACHE_STATUS);
+        return cached == null || !((Boolean) cached.getValue());
```
I would like to see an instanceof check to make sure getValue() can be cast to a Boolean.

rzwiefel commented on catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/CqlCommands.java:
```diff
@@ -0,0 +1,225 @@
+                    .filter(temporalProperty::equalsIgnoreCase)
+                    .findFirst();
+            if (property.isPresent()) {
+                return property.get();
```
or if temporal properties was a Map(name ->Attribute)  the entire method could be done something like
```java
Optional.ofNullable(temporalProperty)
    .filter(StringUtils::isNotEmpty)
    .map(TEMPORAL_PROPERTIES::get)
    .orElse(DEFAULT_TEMPORAL_PROPERTY)
```

glenhein commented on catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/CqlCommands.java:
```diff
@@ -0,0 +1,238 @@
+    protected String getTemporalProperty() {
+        if (metacardTypes != null && StringUtils.isNotEmpty(temporalProperty)) {
+            return metacardTypes.stream()
+                    .map(metacardType -> metacardType.getAttributeDescriptors())
```
Can this be reduced to MetacardType::getAttributeDescriptors?

glenhein commented on catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/CqlCommands.java:
```diff
@@ -0,0 +1,238 @@
+        return DEFAULT_TEMPORAL_PROPERTY;
+    }
+
+    private Predicate<AttributeDescriptor> isDateType() {
```
As an alternative, you could have a method that takes an AttributeDescriptor and returns boolean. The you would use it as: .filter(this::isDateType).

glenhein commented on catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/CqlCommands.java:
```diff
@@ -0,0 +1,238 @@
+
+    private Boolean notNull(String... parameters) {
+        return Arrays.stream(parameters)
+                .anyMatch(p -> p != null);
```
I think you can also use ".anyMatch(Objects::nonNull)"

glenhein commented on catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/CqlCommands.java:
```diff
@@ -0,0 +1,238 @@
+                lastMonths);
+    }
+
+    private Boolean notNull(String... parameters) {
```
It looks like this method will return TRUE if any parameter is non-null. But the method name makes it sound like it return TRUE if all parameters are non-null.

glenhein commented on catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/CqlCommands.java:
```diff
@@ -0,0 +1,238 @@
+                .anyMatch(p -> p != null);
+    }
+
+    private Boolean notZero(int... parameters) {
```
It looks like this method will return TRUE if any parameter is >0. But the method name makes it sound like it return TRUE if all parameters are >0.

peterhuffer commented on catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/SeedCommand.java:
```diff
@@ -69,10 +60,11 @@
                 + "for this command to seed the resource cache.")
 //@formatter:on
-public class SeedCommand extends SubjectCommands implements Action {
+public class SeedCommand extends CqlCommands {
```
Wanted to check with @jrnorth here to make sure we want these additional CqlCommands here. If I remember correctly, there might be a reason some of these Cql options were not added in the first place.

jrnorth replied:
It should be fine to have them. The only issue is that this command won't work with the `--provider` option because it needs to be run through the framework.

peterhuffer commented on catalog/core/catalog-core-commands/src/test/java/org/codice/ddf/commands/catalog/DescribeCommandTest.java:
```diff
@@ -0,0 +1,117 @@
+
+    @Before
+    public void setUp() throws Exception {
+        consoleOutput = new ConsoleOutput();
```
I think we can move this `new ConsoleOutput()` and `.interceptSystemOut` into a `@BeforeClass` method, then call the `consoleOutput.reset()` in the `@After` method and the `.closeBuffer()` in the `@AfterClass` method. Should save us a few object creations. Not sure how necessary the `resetSystemOut` is here if we are going to be intercepting again immediately before each test.

emmberk replied:
I updated the unit tests to extend a common `AbstractConsoleOutputTest` class to include these before/after methods for the consoleOutput.

Lambeaux commented on catalog/core/catalog-core-commands/src/test/java/org/codice/ddf/commands/catalog/IngestCommandTest.java:
```diff
@@ -186,9 +171,8 @@ public void testExpectedCountsWithIgnore() throws Exception {
-    @Ignore
-    //TODO: Use a conditional ignore from JUnit to skip this test on Windows
     public void testIgnoreHiddenFiles() throws Exception {
+        assumeFalse(SystemUtils.IS_OS_WINDOWS);
```
Given there is no Java API to create hidden files, or to set the hidden property for a file on any platform, I would say this is fine. One note: I would make sure that whichever runner is being used, that is `@RunWith(...)`, will honor the `org.junit.Assume` API (http://junit.sourceforge.net/javadoc/org/junit/Assume.html). Might need to check the base class for that annotation. 

See also: 
https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/SystemUtils.html#IS_OS_WINDOWS

http://site.mockito.org/mockito/docs/current/org/mockito/runners/MockitoJUnitRunner.html

Seems like everything checks out 👍

peterhuffer commented on catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/RemoveCommand.java:
```diff
@@ -39,28 +38,21 @@
+@Service
 @Command(scope = CatalogCommands.NAMESPACE, name = "remove", description = "Deletes a record from the Catalog.")
-public class RemoveCommand extends CatalogCommands {
+public class RemoveCommand extends CqlCommands {
```
During heroing and looking at the `--help`, since this command takes a list of `Metacard` IDs as input, I'm not sure it makes any sense to have CqlOptions associated with it.

emmberk commented:
I updated:

- `catalog:remove` to
  - include better descriptions
  - to filter records from only the id list when the id list is set. Before this PR, if a filter and id list was set, all metacards meeting the filter criteria *and* all of the metacards in the id list would be removed.
- the description of the `--temporalProperty` for `CqlCommands` to say "Valid values include*, but are not limited to,* ..." from @peterhuffer 's suggestion
- the description for `catalog:validate` from @peterhuffer 's suggestion
- the descriptions for the options for `CqlCommands` to not include references to ingest or query so that the descriptions are general enough for all of the commands that extend that class
- renamed unit test Abstract... classes to ...Common

emmberk commented on distribution/docs/src/main/resources/_contents/running-contents.adoc:
```diff
@@ -345,101 +345,119 @@ For example, `help search` (see results of this command in the example below).
+        --caseSensitive, -case
+                Option to set the --searchPhrase to be case sensitive.
+        --provider, -p, -provider
+                Interacts with the provider directly instead of the framework.
```
Update description to clarify that this picks _the first_ Catalog provider

peterhuffer commented:
Successfully Herod. 👍 

Verified all —help messages.

Saw expected results for Dump and Search commands, tested —temporalProperty option with created and modified in conjunction with different —lastXXX options.

Remove command: successfully tried —temporalProperty and —lastXXX, and additionally deleting by specifying ID, as well as without specifying ID and only using a  filter (i.e., `catalog:remove —lastMinutes 10`). Also used these in conjunction, and was successful when the filter should have filtered out the metacard ID I specified.

For the Replicate command, I configured a WFS 1.0 source with GeoServer and replicate from the source into the catalog.

Migrate command: installed the `catalog-solr-embedded-provider` feature. And was able to successfully migrate between the two.

Seed command: Setup 2 instances of DDF and federated one to the other and ran the seed command with the -s flag and successfully downloaded metacards and populated cache. Tried successfully with various —lastXXX options as well.
